### PR TITLE
Serialize async updates of user settings.

### PR DESCRIPTION
**Pull request by jimmc**
This PR updates the updateUserSetting function to be truly asynchronous, using promises. If multiple calls are made before previous calls have completed, they are queued up by the simple expedient of chaining them together as promises.

### CODE REVIEW

File: sources/web/datalab/settings.ts [hunk 1]
```diff
@@ -202,47 +205,56 @@ function ensureDirExists(fullPath: string): boolean {
 }
 
 /**
- * Updates a single configuration setting for the user.
+ * Asynchronously updates the user's settings file with the new value for the given key.
+ * If there is already an asynchronous update in progress, this request is queeud up for
```
Comment (yebrahim) [inline sources/web/datalab/settings.ts]: typo: queued

File: sources/web/datalab/settings.ts [hunk 2]
```diff
@@ -202,47 +205,56 @@ function ensureDirExists(fullPath: string): boolean {
...
+      }
+      catch (e) {
+        console.log(e);
+        throw new Error('Failed to read settings');
+      }
+    }
+    if (settings[key] == value) {
+      return false;   // No change was required
+    }
+    settings[key] = value;
```
Comment (yebrahim) [inline sources/web/datalab/settings.ts]: A log message would be useful here.

File: sources/web/datalab/settings.ts [hunk 3]
```diff
@@ -354,16 +366,19 @@ function formHandler(userId: string, formData: any, request: http.ServerRequest,
...
+
+  updateUserSettingAsync(userId, key, value).then(() => {
     if ('redirect' in formData) {
       response.writeHead(302, { 'Location': formData['redirect'] });
     } else {
       response.writeHead(200, { 'Content-Type': 'text/plain' });
     }
-  } else {
+    response.end();
+  }, (errorMessage) => {
```
Comment (yebrahim) [inline sources/web/datalab/settings.ts]: `catch` instead of error callback?

File: sources/web/datalab/settings.ts [hunk 4]
```diff
@@ -202,47 +205,56 @@ function ensureDirExists(fullPath: string): boolean {
...
+      if (this.ensureDirExists(path.normalize(settingsDir))) {
+        fs.writeFileSync(settingsPath, settingsString);
+      }
     }
     catch (e) {
       console.log(e);
-      return false;
+      throw new Error('Failed to write settings');
     }
+    return true;    // File was updated
```
Comment (yebrahim) [inline sources/web/datalab/settings.ts]: Is the boolean return useful for anything or just tests?

**Comment by jimmc** [comment]
Had to make a few little changes to the tests. PTAL.

File: sources/web/datalab/settings.ts (hunk 2 above)
Comment (yebrahim) [inline sources/web/datalab/settings.ts]: Can we add a console log for now then? It's not super important, so up to you.

**Comment by jimmc** [comment]
Added a couple of console.log statements.

**Pull request merged**